tools/common/schema/updatetask_test.go

```diff
@@ -35,57 +37,93 @@ import (
 )
 
 type UpdateTaskTestSuite struct {
-	*require.Assertions // override suite.Suite.Assertions with require.Assertions; this means that s.NotNil(nil) will stop the test, not merely log an error
+	*require.Assertions // override suite.Suite.Assertions with require.Assertions; this means that s.NotNil(n) will stop the test, not merely log an error
 	suite.Suite
+	versionsDir string
+	emptyDir string
+	logger log.Logger
 }
 
 func TestUpdateTaskTestSuite(t *testing.T) {
 	suite.Run(t, new(UpdateTaskTestSuite))
 }
 
+var updateTaskTestData = struct {
+	versions []string
+}{
+	versions: []string{"v0.5", "v1.5", "v2.5", "v2.5.1", "v2.5.2", "v2.7.17", "v3.5", "v10.2", "abc", "2.0", "3.0"},
+}
+
 func (s *UpdateTaskTestSuite) SetupSuite() {
 	s.Assertions = require.New(s.T())
-}
 
-func (s *UpdateTaskTestSuite) TestReadSchemaDir() {
-	emptyDir := testhelper.MkdirTemp(s.T(), "", "update_schema_test_empty")
-	tmpDir := testhelper.MkdirTemp(s.T(), "", "update_schema_test")
+	s.versionsDir = testhelper.MkdirTemp(s.T(), "", "update_schema_test")
 
-	subDirs := []string{"v0.5", "v1.5", "v2.5", "v3.5", "v10.2", "abc", "2.0", "3.0"}
-	for _, d := range subDirs {
-		s.NoError(os.Mkdir(tmpDir+"/"+d, os.FileMode(0444)))
+	for _, d := range updateTaskTestData.versions {
+		s.NoError(os.Mkdir(s.versionsDir+"/"+d, os.FileMode(0444)))
 	}
 
-	_, err := readSchemaDir(tmpDir, "11.0", "11.2")
-	s.Error(err)
-	_, err = readSchemaDir(tmpDir, "0.5", "10.3")
-	s.Error(err)
-	_, err = readSchemaDir(tmpDir, "1.5", "0.5")
-	s.Error(err)
-	_, err = readSchemaDir(tmpDir, "10.3", "")
-	s.Error(err)
-	_, err = readSchemaDir(emptyDir, "11.0", "")
-	s.Error(err)
-	_, err = readSchemaDir(emptyDir, "10.1", "")
-	s.Error(err)
+	s.emptyDir = testhelper.MkdirTemp(s.T(), "", "update_schema_test_empty")
+
+	s.logger = log.NewUnitTestLogger(s.T())
+}
 
-	ans, err := readSchemaDir(tmpDir, "1.5", "1.5")
+func (s *UpdateTaskTestSuite) TestReadSchemaDir() {
+	ans, err := readSchemaDir(s.versionsDir, "10.3", "", s.logger)
 	s.NoError(err)
 	s.Equal(0, len(ans))
 
-	ans, err = readSchemaDir(tmpDir, "0.4", "10.2")
+	ans, err = readSchemaDir(s.versionsDir, "0.5", "1.5", s.logger)
+	s.NoError(err)
+	s.Equal([]string{"v1.5"}, ans)
+
+	ans, err = readSchemaDir(s.versionsDir, "0.4", "10.2", s.logger)
 	s.NoError(err)
-	s.Equal([]string{"v0.5", "v1.5", "v2.5", "v3.5", "v10.2"}, ans)
+	s.Equal([]string{"v0.5", "v1.5", "v2.5", "v2.5.1", "v2.5.2", "v2.7.17", "v3.5", "v10.2"}, ans)
 
-	ans, err = readSchemaDir(tmpDir, "0.5", "3.5")
+	ans, err = readSchemaDir(s.versionsDir, "0.5", "3.5", s.logger)
 	s.NoError(err)
-	s.Equal([]string{"v1.5", "v2.5", "v3.5"}, ans)
+	s.Equal([]string{"v1.5", "v2.5", "v2.5.1", "v2.5.2", "v2.7.17", "v3.5"}, ans)
 
-	ans, err = readSchemaDir(tmpDir, "10.2", "")
+	ans, err = readSchemaDir(s.versionsDir, "10.2", "", s.logger)
 	s.NoError(err)
 	s.Equal(0, len(ans))
```

Comment:
Both test cases with envVer="" return empty result. We need to add one more case to cover envVer="" and result is non-empty, like startVer=v2.5, endVer="".

Reply:
Good catch. Done.